### DISCUSSION TITLE
Inliner and Deforestation Ideas

### DISCUSSION THREAD

**Discussion by magnus-madsen** [flix/flix#8844] [Ideas]
We want two-level inlining and deforestation:

- We know that the JVM JIT will also perform inlining => we should set it up for success. Concretely, this means: 
  - We should ensure that functions are smaller than some k threshold (in bytes).
  - We should ensure that call sites are monomorphic.

Thus, we want to:
- Simplify functions to reduce their bytecount. In particular, if we can bring a function under k bytes that is great.
- Inline and deforest, but without increasing the bytecount above k.
- Not waste time on anything the JIT will do anyway. (E.g. no reason to inline small methods that do not lead to deforestation).
- Possibly break larger functions into smaller ones?

By deforestation I mean both to avoid the allocation of intermediate data structures (e.g. `Option`, `Result`, ...), but also to avoid the allocation of closures.

**Comment by magnus-madsen** [comment]
CC @jaschdoc

**Comment by JonathanStarup** [comment]
> Not waste time on anything the JIT will do anyway. (E.g. no reason to inline small methods that do not lead to deforestation).

Doing inlining that JIT would do anyway could decrease code size - the codesize goal and the goal to not do duplicated JIT work seem to contradict.

> Possibly break larger functions into smaller ones?

This sounds a bit like breaking a `2*k` size function into two, such that JIT can then put it back together?
